clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -278,38 +305,115 @@
+        balance(currentAssignment, prevAssignment, sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions,
                 consumer2AllPotentialPartitions, partition2AllPotentialConsumers, currentPartitionConsumer);
         return currentAssignment;
     }
 
+    private void updatePrevAssignment(Map<TopicPartition, ConsumerGenerationPair> prevAssignment,
+                                      TopicPartition partition, ConsumerGenerationPair consumerGeneration) {
+        if (!prevAssignment.containsKey(partition))
+            prevAssignment.put(partition, consumerGeneration);
+        else {
+            ConsumerGenerationPair existingPair = prevAssignment.get(partition);
+            if (consumerGeneration.generation > existingPair.generation)
+                prevAssignment.put(partition, consumerGeneration);
+            else if (consumerGeneration.generation == existingPair.generation) {
+                // same partition is assigned to two consumers during the same rebalance.
+                // log an error and skip this record
+                log.error("Partition '" + partition + "' had been assigned to multiple consumers " +
+                        "following sticky assignment generation " + consumerGeneration.generation + ".");
+            }
+        }
+    }
+
     private void prepopulateCurrentAssignments(Map<String, Subscription> subscriptions,
-                                               Map<String, List<TopicPartition>> currentAssignment) {
+                                               Map<String, List<TopicPartition>> currentAssignment,
+                                               Map<TopicPartition, ConsumerGenerationPair> prevAssignment) {
+        // we need to process subscriptions' user data with each consumer's reported generation in mind
+        // higher generations overwrite lower generations in case of a conflict
+        // note that a conflict could exists only if user data is for different generations
+
+        // currentPartitionConsumer will hold the latest ConsumerGenerationPair associated with a partition
+        // while prevAssignment holds the prior ConsumerGenerationPair (before current) of each parition
+        Map<TopicPartition, ConsumerGenerationPair> currentPartitionConsumer = new HashMap<>();
         for (Map.Entry<String, Subscription> subscriptionEntry : subscriptions.entrySet()) {
+            String consumer = subscriptionEntry.getKey();
             ByteBuffer userData = subscriptionEntry.getValue().userData();
-            if (userData != null && userData.hasRemaining())
-                currentAssignment.put(subscriptionEntry.getKey(), deserializeTopicPartitionAssignment(userData));
+            if (userData == null || !userData.hasRemaining()) continue;
+            ConsumerUserData consumerUserData = deserializeTopicPartitionAssignment(userData);
+
+            for (TopicPartition partition: consumerUserData.partitions) {
```

Comment:
Effectively what we are doing is sorting the assignments for each partition using the generation and picking the latest two for the current and previous assignments. Is that right? Could we simplify the logic by building a `SortedMap<Integer, String>` for each partition where the key is the generation and the value is the memberID? Then the current assignment would be the last entry and the previous assignment would be the one prior.

Reply:
Great suggestion! It helps simplifying the code and makes `currentPartitionConsumer` redundant. I tried refactoring the code to accommodate this in the new patch.